### PR TITLE
helm/idol-nifi: Scale NiFi statefulset base upon Prometheus metrics

### PR DESCRIPTION
**Pull request by sam-jeffrey-mf**
- Add a PrometheusReportingTask to each NiFi cluster node.
- Add a Prometheus server to the helm deployment
- Report custom metrics from Prometheus to Kubernetes using the PrometheusAdapter 
- Added scaling based upon two metrics: 
    -- average queue utilization percentage
    -- total queued item count

### CODE REVIEW

File: helm/idol-nifi/templates/prometheus/prometheus.yaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# BEGIN COPYRIGHT NOTICE
```
Comment (daniel-lau-mf) [inline helm/idol-nifi/templates/prometheus/prometheus.yaml:1]: Should we make all the `prometheus` bits conditional on `nifi.autoScaling.enabled` (and or some other separate control if it's useful to have the reporting for other purposes but you don't want to run autoscale?)

**Comment by daniel-lau-mf** [comment]
Did you consider https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus ?

File: helm/idol-nifi/values.yaml [hunk 2]
```diff
@@ -217,6 +232,39 @@ nifiRegistry:
       cpu: 1000m
       memory: 1Gi
 
+nifiMetrics:
+  servicePort: 443
+  containerPort: 6443
```
Comment (daniel-lau-mf) [inline helm/idol-nifi/values.yaml:237]: Is there any purpose to these being configurable? As far as I can see both only used within `nifi-metrics.yaml` and I can't see that the ports could clash with anything.

File: helm/idol-nifi/templates/prometheus/ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+metadata:
+  name: nifi-prometheus
+  labels: {{- include "idol-library.labels" (dict "root" . "component" .Values.nifiPrometheus) | nindent 4 -}}
+  annotations:
+{{- if eq .Values.ingressType "nginx" }}
+    nginx.ingress.kubernetes.io/affinity: cookie
+    nginx.ingress.kubernetes.io/affinity-mode: persistent
+{{- end }}
+{{/* ingress setup is very dependent on cluster ingress controller so allow user value annotations */ -}}
+{{ with .Values.nifiRegistry.ingress.annotations }}
```
Comment (daniel-lau-mf) [inline helm/idol-nifi/templates/prometheus/ingress.yaml:23]:
```suggestion
{{ with .Values.nifiPrometheus.ingress.annotations }}
```

File: helm/idol-nifi/values.yaml (hunk 2 above)
Comment (daniel-lau-mf) [reply helm/idol-nifi/values.yaml:237]: If anything needs to be configurable I would argue it's the `9092` port being used on the nifi container to serve the metrics, as presumably the flow could conceivably want to do something else on that port (though this seems unlikely as you'd have to create an additional k8s Service etc if you wanted to access it).

File: helm/idol-nifi/templates/nifi/ingress-metrics.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  labels: {{- include "idol-library.labels" (dict "root" . "component" .Values.nifi) | nindent 4 -}}
+  annotations:
+    app.kubernetes.io/name: nifi-metrics
+    app.kubernetes.io/part-of: nifi
+{{- if eq .Values.ingressType "nginx" }}
+    nginx.ingress.kubernetes.io/affinity: cookie
+    nginx.ingress.kubernetes.io/affinity-mode: persistent
+{{- end }}
+{{/* ingress setup is very dependent on cluster ingress controller so allow user value annotations */ -}}
+{{ with .Values.nifiRegistry.ingress.annotations }}
```
Comment (daniel-lau-mf) [inline helm/idol-nifi/templates/nifi/ingress-metrics.yaml:25]:
```suggestion
{{ with .Values.nifi.ingress.annotations }}
```

File: helm/idol-nifi/templates/nifi/ingress-metrics.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      - path: /metrics
+        pathType: Prefix
+        backend:
+          service:
+            name: idol-nifi
+            port:
+              name: metrics
+{{- if .Values.nifi.ingress.PrefixHost }}
+    host: {{ .Values.nifi.ingress.metricsHost | quote }}
+{{- end }}
```
Comment (daniel-lau-mf) [inline helm/idol-nifi/templates/nifi/ingress-metrics.yaml:50]: Neither of these exist and should presumably be the same thing.

File: helm/idol-nifi/values.yaml (hunk 2 above)
Comment (sam-jeffrey-mf) [reply helm/idol-nifi/values.yaml]: I have removed the configuration options for these and used the values directly in nifi-metrics.yaml.  These were configurable as I was using https://github.com/opentext-idol/idol-containers-toolkit/blob/main/helm/distributed-idol/templates/custom-metrics.yaml#L12-L13 as an example to follow.

You could argue that any/all of the ports used by the nifi container should be configurable as someone may want to use them for something else. As the other ports used by the container are not yet configurable, I think making all of them configurable could be done as a separate change

File: helm/idol-nifi/templates/prometheus/prometheus.yaml (hunk 1 above)
Comment (sam-jeffrey-mf) [reply helm/idol-nifi/templates/prometheus/prometheus.yaml:1]: Added two new enabled flags:
    nifiMetrics.enabled - enables/disables nifi custom metrics
    nifiPrometheus.enabled - enables/disables nifi prometheus monitoring

**Comment by sam-jeffrey-mf** [comment]
> Did you consider https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus ?

Updated to use https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus and https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus-adapter

File: helm/idol-nifi/values.yaml [hunk 6]
```diff
@@ -239,3 +256,55 @@ idol-licenseserver:
...
+      pathType: Prefix
+      path: /prometheus
+    global:
+      scrape_interval: 15s
+
+prometheus-adapter:
+  enabled: true
+  metricsRelistInterval: 30s
+  prometheus:
+    url: http://idol-nifi-prometheus-server
```
Comment (daniel-lau-mf) [inline helm/idol-nifi/values.yaml:278]:
```suggestion
    url: http://{{ .Release.Name }}-prometheus-server
```
Needs to match how the prometheus service will be named. ['Value is templated'](https://github.com/prometheus-community/helm-charts/blob/c45f8fb79f8708250b8a9d0bb417a7abc97229c0/charts/prometheus-adapter/values.yaml#L35) according to the chart docs.

File: helm/idol-nifi/values.yaml [hunk 7]
```diff
@@ -239,3 +256,63 @@ idol-licenseserver:
...
+  alertmanager:
+    enabled: false
+  prometheus-node-exporter:
+    enabled: false
+  prometheus-pushgateway:
+    enabled: false
+  kube-state-metrics:
+    enabled: false
+
+prometheus-adapter:
```
Comment (daniel-lau-mf) [inline helm/idol-nifi/values.yaml:282]:
```suggestion
# -- `prometheus-adapter` sub-chart configuration
# Required for auto-scaling. 
# See https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus-adapter
# @default -- Default configuration to support `idol-nifi` autoscaling. See values.yaml for details.
prometheus-adapter:
```
Suggested docs for generated README

File: helm/idol-nifi/values.yaml [hunk 8]
```diff
@@ -239,3 +256,63 @@ idol-licenseserver:
 metrics-server:
   args:
     - "--kubelet-insecure-tls"
+
+prometheus:
```
Comment (daniel-lau-mf) [inline helm/idol-nifi/values.yaml:260]:
```suggestion
# -- `prometheus` sub-chart configuration
# Required for auto-scaling. 
# See https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus
# @default -- Default configuration to support `idol-nifi` autoscaling. See values.yaml for details.
prometheus:
```
Suggested docs for generated README

File: helm/idol-nifi/values.yaml [hunk 9]
```diff
@@ -97,21 +97,26 @@ nifi:
     minReplicas: 1
     # -- the maximum size of the nifi statefulset
     maxReplicas: 8
+    # -- no. of seconds a limit must be exceed before scaling the nifi statefulset
+    stabilizationWindowSeconds: 300
     # -- one or more metrics controlling the horizontal pod autoscaler (https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v2/#HorizontalPodAutoscalerSpec)
     metrics:
```
Comment (daniel-lau-mf) [inline helm/idol-nifi/values.yaml:103]:
```suggestion
    # -- one or more metrics controlling the horizontal pod autoscaler (https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v2/#HorizontalPodAutoscalerSpec)
    # @default -- See below
    metrics:
```
Slightly nicer for README generation

**Comment by daniel-lau-mf** [comment]
Managed to deploy this on a microk8s test cluster and seemed to scale up/down as described. I had to disable the `metrics-server` microk8s add-on - not sure whether that was a clash with the existing `metrics-server` subchart or the `prometheus-adapter` one.

Maybe useful to add some sort of autoscaling overview in the `README.md.gotmpl` to describe what people need to do in order for NiFi to usefully autoscale (consider where to put Load Balance strategy into connections)?

**Comment by sam-jeffrey-mf** [comment]
> Managed to deploy this on a microk8s test cluster and seemed to scale up/down as described. I had to disable the `metrics-server` microk8s [add-on](https://microk8s.io/docs/addons) - not sure whether that was a clash with the existing `metrics-server` subchart or the `prometheus-adapter` one.

I see this as well if I enable the metrics server addon in my minikube test cluster. This appears to be a clash with the existing metrics-server dependency. If that dependency is removed from the chart.yaml, the chart installs ok.

> Maybe useful to add some sort of autoscaling overview in the `README.md.gotmpl` to describe what people need to do in order for NiFi to usefully autoscale (consider where to put Load Balance strategy into connections)?

Added a section to the README.md.gotmpl around NiFi scaling

**Review by daniel-lau-mf — APPROVED** [review]
Move the `Scaling NiFi` section above the chart values in the README template, but otherwise this looks good now.

**Pull request merged**